File: 0076 - Minimum Window Substring/solution.py
class Solution:
    from collections import Counter
    def minWindow(self, s: str, t: str) -> str:
        m, l, start, L, cnt = Counter(t), 0, -1, float('inf'), len(t)
        for r, c in enumerate(s):
            if m[c]>0:
                cnt -= 1
            m[c] -= 1
            while cnt==0:
                if r-l+1<L:
                    start, L = l, r-l+1
                if m[s[l]] == 0:
                    cnt += 1
                m[s[l]], l = m[s[l]]+1, l+1
        return '' if start==-1 else s[start:start+L]

File: 0076 - Minimum Window Substring/solution.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        vector<int> m(128);
        for(auto c : t) m[c]++;
        int l = 0, r = 0, start = -1, len = INT_MAX, cnt = t.size();
        for(r = 0; r < s.size(); r++){
            if(m[s[r]]-- > 0) cnt--;
            while(cnt == 0){
                if(r-l+1 < len) len = r-l+1, start = l;
                if(m[s[l++]]++ == 0) cnt++;
            }
        }
        return start==-1?"":s.substr(start, len);
    }
};